File: src/main/java/dev/manolovn/validatebst/ValidateBST.java
package dev.manolovn.validatebst;

import dev.manolovn.common.TreeNode;

/**
 * source: https://leetcode.com/problems/validate-binary-search-tree/
 * difficulty: EASY
 * topic: BST
 */
public class ValidateBST {

    /**
     * Time:  O(n)
     * Space: O(n)
     */
    public boolean isValidBST(TreeNode root) {
        return rec(root, null, null);
    }

    private boolean rec(TreeNode n, Integer min, Integer max) {
        if (n == null) return true;
        if (min != null && n.val <= min) return false;
        if (max != null && n.val >= max) return false;
        return rec(n.left, min, n.val) && rec(n.right, n.val, max);
    }
}

File: src/main/kotlin/dev/manolovn/crackingcodinginterview/ch04/ValidateBST.kt
package dev.manolovn.crackingcodinginterview.ch04

/**
 * Implement a function to check if a binary tree is a binary search tree
 */
object ValidateBST {

    fun isBST(root: Node): Boolean = isBSTRec(root, Int.MIN_VALUE, Int.MAX_VALUE)

    private fun isBSTRec(node: Node?, min: Int, max: Int): Boolean {
        if (node == null) return true
        return if (node.value < min || node.value > max) {
            return false
        } else {
            isBSTRec(node.left, min, node.value - 1)
                    && isBSTRec(node.right, node.value + 1, max)
        }
    }

    data class Node(
        val value: Int,
        val left: Node?,
        val right: Node?
    )
}